src/librustc/mir/interpret/value.rs

```diff
@@ -42,6 +43,10 @@ pub enum ConstValue<'tcx> {
     /// An allocation together with a pointer into the allocation.
     /// Invariant: the pointer's `AllocId` resolves to the allocation.
     ByRef(Pointer, &'tcx Allocation),
+
+    /// Used in the HIR by using `Unevaluated` everywhere and later normalizing to one of the other
+    /// variants if the code is monomorphic enough for that.
```

Comment:
IMO this should be at the top with `Param` / `Infer` - but also, I wouldn't mind having `Scalar`, `Slice` and `ByRef` in a sub-enum, because they are the "concrete values".

Maybe keep only `Scalar`, `Slice` and `ByRef` here and move everything else to a `ty::ConstKind` enum?

Reply:
Such a change will cause major fallout and be very bitrotty. I can do that in a separate PR after const generics has been merged as to not block the const generics work.

Reply:
I guess, I was just hoping to avoid having "known value" methods on the general type.

However, IIRC miri uses its own enum for values so this should be fine.

Reply:
opened as https://github.com/rust-lang/rust/issues/59210